File: frontend/js/ventes.js
import { requireAuth, logout } from "./auth.js";
import {
  apiFetch,
  showToast,
  formatDate,
  formatEur,
  tableLoading,
  tableEmpty,
  confirmAction,
  escHtml,
} from "./app.js";

requireAuth();
window.logout = logout;

const STATE = {
  page: 0,
  limit: 50,
  search: "",
  categorie: "",
  plateforme: "",
  total: 0,
  editId: null,
  detailId: null,
  achats: [],
  flips: [],
};

// ── Init ──────────────────────────────────────────────────────────────────────

async function init() {
  setupSidebar();

  const form = document.querySelector('[name="date"]');
  if (form) form.value = new Date().toISOString().split("T")[0];

  await Promise.all([
    loadCategories(),
    loadPlateformes(),
    loadAchats(),
    loadFlips(),
  ]);
  await Promise.all([loadVentes(), loadStats()]);
}

function setupSidebar() {
  const user = JSON.parse(localStorage.getItem("current_user") || "{}");
  const el = document.getElementById("sidebar-username");
  const ini = document.getElementById("sidebar-initial");
  if (el) el.textContent = user.username || "";
  if (ini) ini.textContent = (user.username || "?")[0].toUpperCase();
}

// ── Données annexes ───────────────────────────────────────────────────────────

async function loadCategories() {
  const data = await apiFetch("/api/parametres/categories?type=vente");
  if (!data) return;
  const cats = data.items || data || [];
  const selForm = document.querySelector('form [name="categorie"]');
  const selFilter = document.getElementById("filter-categorie");
  cats.forEach((c) => {
    [selForm, selFilter].forEach((sel) => {
      if (!sel) return;
      const opt = document.createElement("option");
      opt.value = c.nom;
      opt.textContent = c.nom;
      sel.appendChild(opt);
    });
  });
}

async function loadPlateformes() {
  const data = await apiFetch("/api/parametres/plateformes?type=vente");
  if (!data) return;
  const plats = data.items || data || [];
  const selForm = document.querySelector('form [name="plateforme"]');
  const selFilter = document.getElementById("filter-plateforme");
  plats.forEach((p) => {
    [selForm, selFilter].forEach((sel) => {
      if (!sel) return;
      const opt = document.createElement("option");
      opt.value = p.nom;
      opt.textContent = p.nom;
      sel.appendChild(opt);
    });
  });
}

async function loadAchats() {
  const data = await apiFetch("/api/achats?limit=500");
  if (data) {
    STATE.achats = data.items || [];
    const sel = document.getElementById("select-achat");
    STATE.achats.forEach((a) => {
      const opt = document.createElement("option");
      opt.value = a.id;
      opt.textContent = `${a.nom} (${formatEur(a.prix_achat)})`;
      sel.appendChild(opt);
    });
  }
}

async function loadFlips() {
  const data = await apiFetch("/api/flips?limit=500&statut=pret_a_vendre");
  if (data) {
    STATE.flips = data.items || [];
    const sel = document.getElementById("select-flip");
    STATE.flips.forEach((f) => {
      const opt = document.createElement("option");
      opt.value = f.id;
      opt.textContent = `${f.nom} — ${f.marque || ""} ${f.modele || ""}`.trim();
      sel.appendChild(opt);
    });
  }
}

// ── Stats ─────────────────────────────────────────────────────────────────────

async function loadStats() {
  const data = await apiFetch("/api/ventes/stats");
  if (!data) return;

  const ca = document.getElementById("stat-ca");
  const nb = document.getElementById("stat-nb");
  const plats = document.getElementById("stat-plateformes");

  if (ca) ca.textContent = formatEur(data.totaux?.ca_total || 0);
  if (nb) nb.textContent = data.totaux?.nb || 0;

  if (plats && data.par_plateforme?.length) {
    plats.innerHTML = data.par_plateforme
      .map(
        (p) =>
          `<span class="badge badge-outline badge-sm">
            ${escHtml(p.plateforme)} · ${formatEur(p.ca)}
           </span>`
      )
      .join("");
  } else if (plats) {
    plats.innerHTML =
      `<span class="text-base-content/40 text-xs">Aucune vente</span>`;
  }
}

// ── Chargement ventes ─────────────────────────────────────────────────────────

async function loadVentes() {
  const tbody = document.getElementById("ventes-tbody");
  tableLoading(tbody, 8);

  const params = new URLSearchParams({
    skip: STATE.page * STATE.limit,
    limit: STATE.limit,
    search: STATE.search,
    categorie: STATE.categorie,
    plateforme: STATE.plateforme,
  });

  const data = await apiFetch(`/api/ventes?${params}`);
  if (!data) return;

  STATE.total = data.total;
  renderTable(data.items);
  renderPagination();
}

function renderTable(items) {
  const tbody = document.getElementById("ventes-tbody");
  if (!items.length) {
    tableEmpty(tbody, 8, "Aucune vente trouvée");
    return;
  }

  tbody.innerHTML = items
    .map((v) => {
      const margeClass =
        v.marge == null ? "" : v.marge >= 0 ? "text-success" : "text-error";
      const source = v.flip_id
        ? `<span class="badge badge-accent badge-sm">Flip #${v.flip_id}</span>`
        : v.achat_id
          ? `<span class="badge badge-ghost badge-sm">Achat #${v.achat_id}</span>`
          : `<span class="text-base-content/40">—</span>`;

      return `
        <tr class="hover cursor-pointer" onclick="openDetail(${v.id})">
          <td>${formatDate(v.date)}</td>
          <td class="font-medium">${escHtml(v.nom)}</td>
          <td>${escHtml(v.categorie || "—")}</td>
          <td>${escHtml(v.plateforme || "—")}</td>
          <td>${source}</td>
          <td class="text-right font-medium">${formatEur(v.prix_vente)}</td>
          <td class="text-right font-medium ${margeClass}">
            ${v.marge != null ? formatEur(v.marge) : "—"}
          </td>
          <td onclick="event.stopPropagation()">
            <div class="flex gap-1">
              <button class="btn btn-xs btn-ghost"
                onclick="openVenteModal(${v.id})" title="Modifier">✏️</button>
              <button class="btn btn-xs btn-ghost text-error"
                onclick="deleteVente(${v.id})" title="Supprimer">🗑️</button>
            </div>
          </td>
        </tr>
      `;
    })
    .join("");
}

function renderPagination() {
  const el = document.getElementById("pagination");
  const start = STATE.page * STATE.limit + 1;
  const end = Math.min((STATE.page + 1) * STATE.limit, STATE.total);
  const totalPages = Math.ceil(STATE.total / STATE.limit);

  el.innerHTML = `
    <span>${STATE.total > 0 ? `${start}–${end} sur ${STATE.total}` : "Aucun résultat"}</span>
    <div class="flex gap-2">
      <button class="btn btn-sm btn-ghost" ${STATE.page === 0 ? "disabled" : ""}
        onclick="changePage(-1)">← Préc.</button>
      <button class="btn btn-sm btn-ghost"
        ${STATE.page >= totalPages - 1 ? "disabled" : ""}
        onclick="changePage(1)">Suiv. →</button>
    </div>
  `;
}

window.changePage = (dir) => {
  STATE.page = Math.max(0, STATE.page + dir);
  loadVentes();
};

// ── Filtres ───────────────────────────────────────────────────────────────────

let _searchTimer = null;
window.debouncedSearch = () => {
  clearTimeout(_searchTimer);
  _searchTimer = setTimeout(() => {
    STATE.search = document.getElementById("search-input").value;
    STATE.page = 0;
    loadVentes();
  }, 350);
};

window.applyFilters = () => {
  STATE.categorie = document.getElementById("filter-categorie").value;
  STATE.plateforme = document.getElementById("filter-plateforme").value;
  STATE.page = 0;
  loadVentes();
};

window.resetFilters = () => {
  STATE.search = "";
  STATE.categorie = "";
  STATE.plateforme = "";
  STATE.page = 0;
  document.getElementById("search-input").value = "";
  document.getElementById("filter-categorie").value = "";
  document.getElementById("filter-plateforme").value = "";
  loadVentes();
};

// ── Validation source ─────────────────────────────────────────────────────────

window.onSourceChange = () => {
  const achatId = document.getElementById("select-achat").value;
  const flipId = document.getElementById("select-flip").value;
  const warn = document.getElementById("source-warning");
  if (achatId && flipId) {
    warn.classList.remove("hidden");
  } else {
    warn.classList.add("hidden");
  }
};

// ── Modal création/édition ────────────────────────────────────────────────────

window.openVenteModal = async (id = null) => {
  STATE.editId = id;
  const modal = document.getElementById("vente-modal");
  const form = document.getElementById("vente-form");
  const title = document.getElementById("vente-modal-title");

  form.reset();
  document.getElementById("source-warning").classList.add("hidden");
  form.date.value = new Date().toISOString().split("T")[0];

  if (id) {
    title.textContent = "Modifier la vente";
    const vente = await apiFetch(`/api/ventes/${id}`);
    if (!vente) return;

    form.date.value = vente.date?.split("T")[0] || "";
    form.nom.value = vente.nom || "";
    form.prix_vente.value = vente.prix_vente ?? "";
    form.categorie.value = vente.categorie || "";
    form.plateforme.value = vente.plateforme || "";
    form.achat_id.value = vente.achat_id || "";
    form.flip_id.value = vente.flip_id || "";
    form.notes.value = vente.notes || "";
  } else {
    title.textContent = "Nouvelle vente";
  }

  modal.showModal();
};

window.submitVente = async (e) => {
  e.preventDefault();
  const form = e.target;

  const achatId = form.achat_id.value;
  const flipId = form.flip_id.value;

  if (achatId && flipId) {
    showToast("Sélectionnez un achat ou un flip, pas les deux", "error");
    return;
  }

  const payload = {
    date: form.date.value,
    nom: form.nom.value.trim(),
    prix_vente: parseFloat(form.prix_vente.value),
    categorie: form.categorie.value || null,
    plateforme: form.plateforme.value || null,
    achat_id: achatId ? parseInt(achatId, 10) : null,
    flip_id: flipId ? parseInt(flipId, 10) : null,
    notes: form.notes.value.trim() || null,
  };

  const url = STATE.editId ? `/api/ventes/${STATE.editId}` : "/api/ventes";
  const method = STATE.editId ? "PUT" : "POST";

  const result = await apiFetch(url, {
    method,
    body: JSON.stringify(payload),
  });

  if (result) {
    showToast(
      STATE.editId ? "Vente mise à jour" : "Vente enregistrée",
      "success"
    );
    document.getElementById("vente-modal").close();
    await Promise.all([loadVentes(), loadStats()]);
  }
};

// ── Suppression ───────────────────────────────────────────────────────────────

window.deleteVente = async (id) => {
  if (!confirmAction("Supprimer cette vente ?")) return;
  const res = await apiFetch(`/api/ventes/${id}`, { method: "DELETE" });
  if (res !== null) {
    showToast("Vente supprimée", "success");
    await Promise.all([loadVentes(), loadStats()]);
  }
};

// ── Modal détail ──────────────────────────────────────────────────────────────

window.openDetail = async (id) => {
  STATE.detailId = id;
  const vente = await apiFetch(`/api/ventes/${id}`);
  if (!vente) return;

  document.getElementById("detail-vente-title").textContent =
    `Vente — ${vente.nom}`;

  const margeClass =
    vente.marge == null
      ? ""
      : vente.marge >= 0
        ? "text-success"
        : "text-error";

  const source = vente.flip_id
    ? `<a href="/flips.html" class="link link-primary">
         Flip #${vente.flip_id} — ${escHtml(vente.flip_nom || "")}
       </a>`
    : vente.achat_id
      ? `<a href="/achats.html" class="link link-primary">
           Achat #${vente.achat_id} — ${escHtml(vente.achat_nom || "")}
         </a>`
      : "—";

  // Fichiers
  const fichiersHtml =
    vente.fichiers?.length
      ? `<div class="flex flex-wrap gap-2">
          ${vente.fichiers
            .map((f) => {
              const isImg = f.mime_type?.startsWith("image/");
              return isImg
                ? `<div class="relative group">
                    <img src="/uploads/${f.chemin}"
                         class="h-20 w-20 object-cover rounded-lg" />
                    <button
                      class="absolute top-1 right-1 btn btn-xs btn-error
                             opacity-0 group-hover:opacity-100 transition-opacity"
                      onclick="deleteFichier(${f.id}, ${id})">✕</button>
                  </div>`
                : `<a href="/uploads/${f.chemin}" target="_blank"
                      class="btn btn-xs btn-ghost">
                     📎 ${escHtml(f.nom_original)}
                   </a>`;
            })
            .join("")}
        </div>`
      : `<p class="text-sm text-base-content/50">Aucun fichier</p>`;

  document.getElementById("detail-vente-body").innerHTML = `
    <div class="stats stats-horizontal shadow w-full">
      <div class="stat place-items-center p-3">
        <div class="stat-title text-xs">Prix vente</div>
        <div class="stat-value text-base font-bold">
          ${formatEur(vente.prix_vente)}
        </div>
      </div>
      <div class="stat place-items-center p-3">
        <div class="stat-title text-xs">Marge</div>
        <div class="stat-value text-base font-bold ${margeClass}">
          ${vente.marge != null ? formatEur(vente.marge) : "—"}
        </div>
      </div>
    </div>

    <div class="grid grid-cols-2 gap-3 text-sm">
      <div>
        <div class="text-xs text-base-content/50 uppercase tracking-wide">Date</div>
        <div>${formatDate(vente.date)}</div>
      </div>
      <div>
        <div class="text-xs text-base-content/50 uppercase tracking-wide">Plateforme</div>
        <div>${escHtml(vente.plateforme || "—")}</div>
      </div>
      <div>
        <div class="text-xs text-base-content/50 uppercase tracking-wide">Catégorie</div>
        <div>${escHtml(vente.categorie || "—")}</div>
      </div>
      <div>
        <div class="text-xs text-base-content/50 uppercase tracking-wide">Source</div>
        <div>${source}</div>
      </div>
    </div>

    ${
      vente.notes
        ? `<div>
            <div class="text-xs text-base-content/50 uppercase tracking-wide mb-1">Notes</div>
            <p class="text-sm text-base-content/70">${escHtml(vente.notes)}</p>
          </div>`
        : ""
    }

    <div>
      <div class="text-xs text-base-content/50 uppercase tracking-wide mb-2">
        Fichiers joints
      </div>
      ${fichiersHtml}
      <div class="mt-3 flex gap-2 items-center">
        <input type="file" id="fichier-input"
               class="file-input file-input-bordered file-input-sm flex-1"
               accept="image/*,.pdf,.csv,.xlsx" />
        <button class="btn btn-sm btn-secondary"
                onclick="uploadFichier(${id})">
          Joindre
        </button>
      </div>
    </div>
  `;

  document.getElementById("vente-detail-modal").showModal();
};

// ── Fichiers ──────────────────────────────────────────────────────────────────

window.uploadFichier = async (venteId) => {
  const input = document.getElementById("fichier-input");
  if (!input.files.length) {
    showToast("Sélectionnez un fichier", "warning");
    return;
  }
  const form = new FormData();
  form.append("file", input.files[0]);

  const result = await apiFetch(`/api/ventes/${venteId}/fichiers`, {
    method: "POST",
    body: form,
  });
  if (result) {
    showToast("Fichier joint", "success");
    input.value = "";
    await openDetail(venteId);
  }
};

window.deleteFichier = async (fichierId, venteId) => {
  if (!confirmAction("Supprimer ce fichier ?")) return;
  const res = await apiFetch(
    `/api/ventes/${venteId}/fichiers/${fichierId}`,
    { method: "DELETE" }
  );
  if (res !== null) {
    showToast("Fichier supprimé", "success");
    await openDetail(venteId);
  }
};

// ── Bootstrap ─────────────────────────────────────────────────────────────────

init();

File: backend/routers/ventes.py
from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile, File
from backend.database import get_db
from backend.auth import get_current_user, require_role
from backend.models import VenteCreate, VenteUpdate
from backend.services.files_service import save_upload, delete_file
import aiosqlite
import json

router = APIRouter(prefix="/api/ventes", tags=["ventes"])


# ── Liste ─────────────────────────────────────────────────────────────────────

@router.get("")
async def list_ventes(
    search: str = Query(default=""),
    categorie: str = Query(default=""),
    plateforme: str = Query(default=""),
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    conditions = []
    params = []

    if search:
        conditions.append("(v.nom LIKE ? OR v.notes LIKE ?)")
        like = f"%{search}%"
        params.extend([like, like])

    if categorie:
        conditions.append("v.categorie = ?")
        params.append(categorie)

    if plateforme:
        conditions.append("v.plateforme = ?")
        params.append(plateforme)

    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""

    total_row = await db.execute_fetchall(
        f"SELECT COUNT(*) as n FROM ventes v {where}", params
    )
    total = total_row[0]["n"]

    rows = await db.execute_fetchall(
        f"""
        SELECT v.*,
               a.nom as achat_nom, a.prix_achat as achat_prix,
               f.nom as flip_nom, f.prix_achat as flip_prix_achat,
               f.cout_pieces as flip_cout_pieces
        FROM ventes v
        LEFT JOIN achats a ON a.id = v.achat_id
        LEFT JOIN flips f ON f.id = v.flip_id
        {where}
        ORDER BY v.date DESC
        LIMIT ? OFFSET ?
        """,
        params + [limit, skip],
    )

    items = []
    for r in rows:
        d = dict(r)
        # Calcul marge selon source
        if d.get("flip_id") and d.get("flip_prix_achat") is not None:
            cout = (d["flip_prix_achat"] or 0) + (d["flip_cout_pieces"] or 0)
            d["marge"] = d["prix_vente"] - cout
        elif d.get("achat_id") and d.get("achat_prix") is not None:
            d["marge"] = d["prix_vente"] - d["achat_prix"]
        else:
            d["marge"] = None
        items.append(d)

    return {"total": total, "items": items}


# ── Stats ─────────────────────────────────────────────────────────────────────

@router.get("/stats")
async def get_stats(
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    par_plateforme = await db.execute_fetchall(
        """
        SELECT plateforme,
               COUNT(*) as nb,
               SUM(prix_vente) as ca
        FROM ventes
        WHERE plateforme IS NOT NULL AND plateforme != ''
        GROUP BY plateforme
        ORDER BY ca DESC
        """,
        [],
    )

    par_categorie = await db.execute_fetchall(
        """
        SELECT categorie,
               COUNT(*) as nb,
               SUM(prix_vente) as ca
        FROM ventes
        WHERE categorie IS NOT NULL AND categorie != ''
        GROUP BY categorie
        ORDER BY ca DESC
        """,
        [],
    )

    totaux = await db.execute_fetchall(
        """
        SELECT COUNT(*) as nb,
               COALESCE(SUM(prix_vente), 0) as ca_total
        FROM ventes
        """,
        [],
    )

    return {
        "totaux": dict(totaux[0]) if totaux else {},
        "par_plateforme": [dict(r) for r in par_plateforme],
        "par_categorie": [dict(r) for r in par_categorie],
    }


# ── Détail ────────────────────────────────────────────────────────────────────

@router.get("/{vente_id}")
async def get_vente(
    vente_id: int,
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    rows = await db.execute_fetchall(
        """
        SELECT v.*,
               a.nom as achat_nom, a.prix_achat as achat_prix,
               f.nom as flip_nom, f.prix_achat as flip_prix_achat,
               f.cout_pieces as flip_cout_pieces
        FROM ventes v
        LEFT JOIN achats a ON a.id = v.achat_id
        LEFT JOIN flips f ON f.id = v.flip_id
        WHERE v.id = ?
        """,
        [vente_id],
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Vente introuvable")

    fichiers = await db.execute_fetchall(
        """
        SELECT * FROM fichiers
        WHERE type_parent = 'vente' AND parent_id = ?
        ORDER BY created_at ASC
        """,
        [vente_id],
    )

    d = dict(rows[0])
    if d.get("flip_id") and d.get("flip_prix_achat") is not None:
        cout = (d["flip_prix_achat"] or 0) + (d["flip_cout_pieces"] or 0)
        d["marge"] = d["prix_vente"] - cout
    elif d.get("achat_id") and d.get("achat_prix") is not None:
        d["marge"] = d["prix_vente"] - d["achat_prix"]
    else:
        d["marge"] = None

    d["fichiers"] = [dict(f) for f in fichiers]
    return d


# ── Création ──────────────────────────────────────────────────────────────────

@router.post("", status_code=201)
async def create_vente(
    payload: VenteCreate,
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    # Si lié à un flip → mettre à jour flip.vente_id + statut vendu
    if payload.flip_id:
        flip_rows = await db.execute_fetchall(
            "SELECT id, statut FROM flips WHERE id = ?", [payload.flip_id]
        )
        if not flip_rows:
            raise HTTPException(status_code=404, detail="Flip introuvable")

    cur = await db.execute(
        """
        INSERT INTO ventes (date, nom, categorie, plateforme,
                            prix_vente, achat_id, flip_id, notes)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """,
        [
            payload.date,
            payload.nom,
            payload.categorie,
            payload.plateforme,
            payload.prix_vente,
            payload.achat_id,
            payload.flip_id,
            payload.notes,
        ],
    )
    await db.commit()
    vente_id = cur.lastrowid

    # Lier le flip à cette vente et passer en "vendu"
    if payload.flip_id:
        await db.execute(
            """
            UPDATE flips
            SET vente_id = ?, statut = 'vendu',
                updated_at = CURRENT_TIMESTAMP
            WHERE id = ?
            """,
            [vente_id, payload.flip_id],
        )

    await db.execute(
        """
        INSERT INTO logs (utilisateur_id, action, entite, entite_id, details)
        VALUES (?, 'create', 'vente', ?, ?)
        """,
        [current_user["id"], vente_id, json.dumps({"nom": payload.nom})],
    )
    await db.commit()

    rows = await db.execute_fetchall(
        "SELECT * FROM ventes WHERE id = ?", [vente_id]
    )
    return dict(rows[0])


# ── Mise à jour ───────────────────────────────────────────────────────────────

@router.put("/{vente_id}")
async def update_vente(
    vente_id: int,
    payload: VenteUpdate,
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    rows = await db.execute_fetchall(
        "SELECT * FROM ventes WHERE id = ?", [vente_id]
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Vente introuvable")

    data = payload.model_dump(exclude_none=True)
    if not data:
        raise HTTPException(status_code=400, detail="Aucun champ à mettre à jour")

    sets = ", ".join(f"{k} = ?" for k in data)
    vals = list(data.values()) + [vente_id]
    await db.execute(
        f"UPDATE ventes SET {sets}, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
        vals,
    )
    await db.execute(
        """
        INSERT INTO logs (utilisateur_id, action, entite, entite_id, details)
        VALUES (?, 'update', 'vente', ?, ?)
        """,
        [current_user["id"], vente_id, json.dumps(data)],
    )
    await db.commit()

    rows = await db.execute_fetchall(
        "SELECT * FROM ventes WHERE id = ?", [vente_id]
    )
    return dict(rows[0])


# ── Suppression ───────────────────────────────────────────────────────────────

@router.delete("/{vente_id}", status_code=204)
async def delete_vente(
    vente_id: int,
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(require_role("admin")),
):
    rows = await db.execute_fetchall(
        "SELECT * FROM ventes WHERE id = ?", [vente_id]
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Vente introuvable")

    vente = dict(rows[0])

    # Délier le flip si besoin
    if vente.get("flip_id"):
        await db.execute(
            """
            UPDATE flips
            SET vente_id = NULL, statut = 'pret_a_vendre',
                updated_at = CURRENT_TIMESTAMP
            WHERE id = ? AND vente_id = ?
            """,
            [vente["flip_id"], vente_id],
        )

    fichiers = await db.execute_fetchall(
        "SELECT chemin FROM fichiers WHERE type_parent = 'vente' AND parent_id = ?",
        [vente_id],
    )
    for f in fichiers:
        delete_file(f["chemin"])

    await db.execute(
        "DELETE FROM fichiers WHERE type_parent = 'vente' AND parent_id = ?",
        [vente_id],
    )
    await db.execute("DELETE FROM ventes WHERE id = ?", [vente_id])
    await db.execute(
        """
        INSERT INTO logs (utilisateur_id, action, entite, entite_id, details)
        VALUES (?, 'delete', 'vente', ?, ?)
        """,
        [current_user["id"], vente_id, json.dumps({"vente_id": vente_id})],
    )
    await db.commit()


# ── Fichiers ──────────────────────────────────────────────────────────────────

@router.post("/{vente_id}/fichiers", status_code=201)
async def upload_fichier(
    vente_id: int,
    file: UploadFile = File(...),
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    rows = await db.execute_fetchall(
        "SELECT id FROM ventes WHERE id = ?", [vente_id]
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Vente introuvable")

    saved = await save_upload(file, "vente", vente_id)

    cur = await db.execute(
        """
        INSERT INTO fichiers
            (type_parent, parent_id, nom_original, chemin, mime_type, taille, categorie)
        VALUES ('vente', ?, ?, ?, ?, ?, 'justificatif')
        """,
        [
            vente_id,
            saved["nom_original"],
            saved["chemin"],
            saved["mime_type"],
            saved["taille"],
        ],
    )
    await db.commit()

    rows = await db.execute_fetchall(
        "SELECT * FROM fichiers WHERE id = ?", [cur.lastrowid]
    )
    return dict(rows[0])


@router.delete("/{vente_id}/fichiers/{fichier_id}", status_code=204)
async def delete_fichier(
    vente_id: int,
    fichier_id: int,
    db: aiosqlite.Connection = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    rows = await db.execute_fetchall(
        """
        SELECT * FROM fichiers
        WHERE id = ? AND type_parent = 'vente' AND parent_id = ?
        """,
        [fichier_id, vente_id],
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Fichier introuvable")

    delete_file(rows[0]["chemin"])
    await db.execute("DELETE FROM fichiers WHERE id = ?", [fichier_id])
    await db.commit()